src/cluster_legacy.c

```diff
@@ -3989,28 +4146,64 @@ int clusterSendModuleMessageToTarget(const char *target,
  * Otherwise:
  * Publish this message across the slot (primary/replica).
  * -------------------------------------------------------------------------- */
-void clusterPropagatePublish(robj *channel, robj *message, int sharded) {
-    clusterMsgSendBlock *msgblock;
-
+void clusterPropagatePublish(robj *channel, robj **messages, int count, int sharded) {
+    clusterMsgSendBlock *msgblock, *msgblock_light;
+    int i;
+    msgblock_light = clusterCreatePublishLightMsgBlock(
+        channel, messages, count, sharded ? CLUSTERMSG_TYPE_PUBLISHSHARD_LIGHT : CLUSTERMSG_TYPE_PUBLISH_LIGHT);
     if (!sharded) {
-        msgblock = clusterCreatePublishMsgBlock(channel, message, CLUSTERMSG_TYPE_PUBLISH);
-        clusterBroadcastMessage(msgblock);
-        clusterMsgSendBlockDecrRefCount(msgblock);
+        clusterBroadcastMessage(msgblock_light);
+        clusterMsgSendBlockDecrRefCount(msgblock_light);
+        if (!server.cluster->is_light_hdr_supported) {
+            for (i = 0; i < count; i++) {
+                msgblock = clusterCreatePublishMsgBlock(channel, messages[i], CLUSTERMSG_TYPE_PUBLISH);
+                clusterBroadcastMessage(msgblock);
+                clusterMsgSendBlockDecrRefCount(msgblock);
+            }
+        }
         return;
     }
 
     listIter li;
     listNode *ln;
+    clusterNode *node;
     list *nodes_for_slot = clusterGetNodesInMyShard(server.cluster->myself);
     serverAssert(nodes_for_slot != NULL);
     listRewind(nodes_for_slot, &li);
-    msgblock = clusterCreatePublishMsgBlock(channel, message, CLUSTERMSG_TYPE_PUBLISHSHARD);
+    int nodes_not_supporting_light_header = 0;
     while ((ln = listNext(&li))) {
-        clusterNode *node = listNodeValue(ln);
+        node = listNodeValue(ln);
         if (node->flags & (CLUSTER_NODE_MYSELF | CLUSTER_NODE_HANDSHAKE)) continue;
-        clusterSendMessage(node->link, msgblock);
+        if (nodeSupportsLightMsgHdr(node)) {
+            clusterSendMessage(node->link, msgblock_light);
+        } else {
+            nodes_not_supporting_light_header++;
+            continue;
+        }
+    }
+    clusterMsgSendBlockDecrRefCount(msgblock_light);
+
+    if (nodes_not_supporting_light_header) {
+        server.cluster->is_light_hdr_supported = 0;
+    } else {
+        server.cluster->is_light_hdr_supported = 1;
+    }
+
+    if (!server.cluster->is_light_hdr_supported) {
+        for (i = 0; i < count; i++) {
+            listRewind(nodes_for_slot, &li);
+            msgblock = clusterCreatePublishMsgBlock(channel, messages[i], CLUSTERMSG_TYPE_PUBLISHSHARD);
+            while ((ln = listNext(&li))) {
```

Comment:
My take would be we should send all the message to a single node and then move to the next message.

Reply:
On hindsight, with this approach `msgblock` needs to be created once whereas with the approach I suggested you need to create it for each node.